include/tuvx/linear_algebra/linear_algebra.inl

```diff
@@ -0,0 +1,99 @@
+
+namespace tuvx {
+
+template <typename T>
+inline TridiagonalMatrix<T>::TridiagonalMatrix(std::size_t size) {
+  this->size_ = size;
+  this->main_diagonal_ = std::vector<T>(size);
+  this->upper_diagonal_ = std::vector<T>(size - 1);
+  this->lower_diagonal_ = std::vector<T>(size - 1);
+}
+
+template <typename T>
+inline std::vector<T> Dot(const TridiagonalMatrix<T> &A,
+                          const std::vector<T> &x) {
+  std::size_t size = x.size();
+  std::vector<T> v(size);
+  v[0] = A.main_diagonal_[0] * x[0] + A.upper_diagonal_[0] * x[1];
+
+  std::size_t i = 0;
+  for (i = 1; i < size - 1; i++) {
+    v[i] = A.main_diagonal_[i] * x[i] + A.upper_diagonal_[i] * x[i + 1] +
+           A.lower_diagonal_[i - 1] * x[i - 1];
+  }
+  v[i] = A.main_diagonal_[i] * x[i] + A.lower_diagonal_[i - 1] * x[i - 1];
+  return v;
+}
+
+template <typename T>
+inline std::vector<T> Solve(TridiagonalMatrix<T> A, std::vector<T> b) {
+  T temp;
+  std::size_t N = b.size();
+  std::vector<T> x(N);
+  // forward pass
+  for (std::size_t i = 1; i < N; i++) {
+    temp = A.lower_diagonal_[i - 1] / A.main_diagonal_[i - 1];
+    A.main_diagonal_[i] -= temp * A.upper_diagonal_[i - 1];
+    b[i] -= temp * b[i - 1];
+  }
+  // back substitution
+  x[N - 1] = b[N - 1] / A.main_diagonal_[N - 1];
+  for (std::size_t i = N - 2;; i--) {
+    x[i] = (b[i] - A.upper_diagonal_[i] * x[i + 1]) / A.main_diagonal_[i];
+    if (i == 0) {
+      break;
+    }
+  }
+  return x;
+}
+
+template <typename T> inline void FillRandom(std::vector<T> &x) {
+  std::random_device dev;
+  std::mt19937 rng(dev());
+  std::uniform_int_distribution<std::mt19937::result_type> dist6(1, 6);
+  for (std::size_t i = 0; i < x.size(); i++) {
+    x[i] = (T)dist6(rng) + 1;
+  }
+}
+
+template <typename T> inline void FillRandom(TridiagonalMatrix<T> &A) {
+  FillRandom<T>(A.main_diagonal_);
+  FillRandom<T>(A.lower_diagonal_);
+  FillRandom<T>(A.upper_diagonal_);
+  // make diagonally dominant (diagonal value greater than sum of its row)
+  std::size_t i = 0;
+  A.main_diagonal_[i] += A.upper_diagonal_[i];
+  for (i = 1; i < A.size_ - 1; i++) {
+    A.main_diagonal_[i] += A.lower_diagonal_[i - 1] + A.upper_diagonal_[i];
+  }
+  A.main_diagonal_[i] += A.lower_diagonal_[i - 1];
+}
+
+template <typename T> inline void Print(const std::vector<T> &x) {
+  std::cout << std::endl;
+  for (std::size_t i = 0; i < x.size(); i++) {
+    std::cout << x.at(i) << std::endl;
+  }
+  std::cout << std::endl;
+}
+
+template <typename T>
+T ComputeError(const std::vector<T> &x, const std::vector<T> &x_approx,
+               int norm_order) {
+  T error = 0;
+  for (std::size_t i = 0; i < x.size(); i++) {
+    error += std::pow(std::abs(x[i] - x_approx[i]), norm_order);
+  }
+  error = ((T)1 / x.size()) * std::pow(error, (T)1 / (T)norm_order);
```

Comment:
Should this be `error = std::pow(error / x.size(), (T)1 / (T)norm_order)`? I am thinking RMSE may be a better measurement than the L2 norm here. What do you think @mwaxmonsky ?

Also we can compute the relative error (i.e., `sum(abs(x[i] - x_approx[i])/max(x[i], x_approx[i]))/x.size()`) and compare it to the machine epsilon.

Reply:
@sjsprecious I agree with using RMSE here.  Unless there is an integration issue with the rest of the tuv-x code that depends on the LP norm, I think RMSE makes more sense in this context.

Reply:
@sjsprecious I am not familiar with the relative error code you shared. Isn't the relative error defined as follows?
$$\frac{||x_{true} - x_{approx}||_p}{ ||x_{true}||_p}$$

Reply:
@sjsprecious @mwaxmonsky I have used the relative error instead of absolute in my latest commit. I have also used the error tolerance to be the machine epsilon instead of setting it arbitrarily. Lastly, I added a test checking the consistency in error over various system sizes.